File: sdks/java/src/main/java/com/puxbay/resources/AdditionalResources.java
package com.puxbay.resources;

import com.puxbay.Puxbay;
import com.puxbay.exceptions.PuxbayException;
import com.puxbay.models.*;
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Inventory resource - Inventory tracking operations.
 */
public class InventoryResource extends BaseResource {
    public InventoryResource(Puxbay client) {
        super(client);
    }
    
    public List<StockLevel> getStockLevels(String branchId) throws PuxbayException {
        Type type = new TypeToken<List<StockLevel>>(){}.getType();
        return client.request("GET", String.format("inventory/stock-levels/?branch=%s", branchId), null, type);
    }
    
    public StockLevel getProductStock(String productId, String branchId) throws PuxbayException {
        return get(String.format("inventory/product-stock/?product=%s&branch=%s", productId, branchId), StockLevel.class);
    }
}

/**
 * Reports resource - Sales reports and analytics.
 */
public class ReportsResource extends BaseResource {
    public ReportsResource(Puxbay client) {
        super(client);
    }
    
    public FinancialSummary getFinancialSummary(String startDate, String endDate) throws PuxbayException {
        return get(String.format("reports/financial-summary/?start_date=%s&end_date=%s", startDate, endDate), FinancialSummary.class);
    }
    
    public List<DailySales> getDailySales(String startDate, String endDate) throws PuxbayException {
        Type type = new TypeToken<List<DailySales>>(){}.getType();
        return client.request("GET", String.format("reports/daily-sales/?start_date=%s&end_date=%s", startDate, endDate), null, type);
    }
    
    public List<Product> getTopProducts(int limit) throws PuxbayException {
        Type type = new TypeToken<List<Product>>(){}.getType();
        return client.request("GET", String.format("reports/top-products/?limit=%d", limit), null, type);
    }
    
    public List<Product> getLowStock() throws PuxbayException {
        Type type = new TypeToken<List<Product>>(){}.getType();
        return client.request("GET", "reports/low-stock/", null, type);
    }
}

/**
 * Purchase Orders resource - Purchase order management.
 */
public class PurchaseOrdersResource extends BaseResource {
    public PurchaseOrdersResource(Puxbay client) {
        super(client);
    }
    
    public PaginatedResponse<PurchaseOrder> list(int page, String status) throws PuxbayException {
        String endpoint = String.format("purchase-orders/?page=%d", page);
        if (status != null && !status.isEmpty()) {
            endpoint += "&status=" + status;
        }
        Type type = new TypeToken<PaginatedResponse<PurchaseOrder>>(){}.getType();
        return client.request("GET", endpoint, null, type);
    }
    
    public PurchaseOrder get(String poId) throws PuxbayException {
        return get(String.format("purchase-orders/%s/", poId), PurchaseOrder.class);
    }
    
    public PurchaseOrder create(PurchaseOrder po) throws PuxbayException {
        return post("purchase-orders/", po, PurchaseOrder.class);
    }
    
    public PurchaseOrder update(String poId, PurchaseOrder po) throws PuxbayException {
        return patch(String.format("purchase-orders/%s/", poId), po, PurchaseOrder.class);
    }
    
    public PurchaseOrder receive(String poId, List<PurchaseOrderItem> items) throws PuxbayException {
        Map<String, Object> body = new HashMap<>();
        body.put("items", items);
        return post(String.format("purchase-orders/%s/receive/", poId), body, PurchaseOrder.class);
    }
}

/**
 * Stock Transfers resource - Stock transfer management.
 */
public class StockTransfersResource extends BaseResource {
    public StockTransfersResource(Puxbay client) {
        super(client);
    }
    
    public PaginatedResponse<StockTransfer> list(int page, String status) throws PuxbayException {
        String endpoint = String.format("stock-transfers/?page=%d", page);
        if (status != null && !status.isEmpty()) {
            endpoint += "&status=" + status;
        }
        Type type = new TypeToken<PaginatedResponse<StockTransfer>>(){}.getType();
        return client.request("GET", endpoint, null, type);
    }
    
    public StockTransfer get(String transferId) throws PuxbayException {
        return get(String.format("stock-transfers/%s/", transferId), StockTransfer.class);
    }
    
    public StockTransfer create(StockTransfer transfer) throws PuxbayException {
        return post("stock-transfers/", transfer, StockTransfer.class);
    }
    
    public StockTransfer complete(String transferId) throws PuxbayException {
        return post(String.format("stock-transfers/%s/complete/", transferId), null, StockTransfer.class);
    }
}

/**
 * Stocktakes resource - Stocktake session management.
 */
public class StocktakesResource extends BaseResource {
    public StocktakesResource(Puxbay client) {
        super(client);
    }
    
    public PaginatedResponse<StocktakeSession> list(int page) throws PuxbayException {
        Type type = new TypeToken<PaginatedResponse<StocktakeSession>>(){}.getType();
        return client.request("GET", String.format("stocktakes/?page=%d", page), null, type);
    }
    
    public StocktakeSession get(String stocktakeId) throws PuxbayException {
        return get(String.format("stocktakes/%s/", stocktakeId), StocktakeSession.class);
    }
    
    public StocktakeSession create(StocktakeSession stocktake) throws PuxbayException {
        return post("stocktakes/", stocktake, StocktakeSession.class);
    }
    
    public StocktakeSession complete(String stocktakeId) throws PuxbayException {
        return post(String.format("stocktakes/%s/complete/", stocktakeId), null, StocktakeSession.class);
    }
}

/**
 * Cash Drawers resource - Cash drawer session management.
 */
public class CashDrawersResource extends BaseResource {
    public CashDrawersResource(Puxbay client) {
        super(client);
    }
    
    public PaginatedResponse<CashDrawerSession> list(int page) throws PuxbayException {
        Type type = new TypeToken<PaginatedResponse<CashDrawerSession>>(){}.getType();
        return client.request("GET", String.format("cash-drawers/?page=%d", page), null, type);
    }
    
    public CashDrawerSession get(String drawerId) throws PuxbayException {
        return get(String.format("cash-drawers/%s/", drawerId), CashDrawerSession.class);
    }
    
    public CashDrawerSession open(CashDrawerSession drawer) throws PuxbayException {
        return post("cash-drawers/", drawer, CashDrawerSession.class);
    }
    
    public CashDrawerSession close(String drawerId, double actualCash) throws PuxbayException {
        Map<String, Object> body = new HashMap<>();
        body.put("actual_cash", actualCash);
        return post(String.format("cash-drawers/%s/close/", drawerId), body, CashDrawerSession.class);
    }
}

/**
 * Expenses resource - Expense tracking operations.
 */
public class ExpensesResource extends BaseResource {
    public ExpensesResource(Puxbay client) {
        super(client);
    }
    
    public PaginatedResponse<Expense> list(int page, String category) throws PuxbayException {
        String endpoint = String.format("expenses/?page=%d", page);
        if (category != null && !category.isEmpty()) {
            endpoint += "&category=" + category;
        }
        Type type = new TypeToken<PaginatedResponse<Expense>>(){}.getType();
        return client.request("GET", endpoint, null, type);
    }
    
    public Expense get(String expenseId) throws PuxbayException {
        return get(String.format("expenses/%s/", expenseId), Expense.class);
    }
    
    public Expense create(Expense expense) throws PuxbayException {
        return post("expenses/", expense, Expense.class);
    }
    
    public Expense update(String expenseId, Expense expense) throws PuxbayException {
        return patch(String.format("expenses/%s/", expenseId), expense, Expense.class);
    }
    
    public void delete(String expenseId) throws PuxbayException {
        delete(String.format("expenses/%s/", expenseId));
    }
    
    public List<ExpenseCategory> listCategories() throws PuxbayException {
        Type type = new TypeToken<List<ExpenseCategory>>(){}.getType();
        return client.request("GET", "expense-categories/", null, type);
    }
}

/**
 * Notifications resource - Notification management.
 */
public class NotificationsResource extends BaseResource {
    public NotificationsResource(Puxbay client) {
        super(client);
    }
    
    public PaginatedResponse<Notification> list(int page) throws PuxbayException {
        Type type = new TypeToken<PaginatedResponse<Notification>>(){}.getType();
        return client.request("GET", String.format("notifications/?page=%d", page), null, type);
    }
    
    public Notification get(String notificationId) throws PuxbayException {
        return get(String.format("notifications/%s/", notificationId), Notification.class);
    }
    
    public Notification markAsRead(String notificationId) throws PuxbayException {
        return post(String.format("notifications/%s/mark-read/", notificationId), null, Notification.class);
    }
}

/**
 * Returns resource - Return processing operations.
 */
public class ReturnsResource extends BaseResource {
    public ReturnsResource(Puxbay client) {
        super(client);
    }
    
    public PaginatedResponse<Return> list(int page) throws PuxbayException {
        Type type = new TypeToken<PaginatedResponse<Return>>(){}.getType();
        return client.request("GET", String.format("returns/?page=%d", page), null, type);
    }
    
    public Return get(String returnId) throws PuxbayException {
        return get(String.format("returns/%s/", returnId), Return.class);
    }
    
    public Return create(Return returnObj) throws PuxbayException {
        return post("returns/", returnObj, Return.class);
    }
    
    public Return approve(String returnId) throws PuxbayException {
        return post(String.format("returns/%s/approve/", returnId), null, Return.class);
    }
}

File: sdks/dotnet/src/Resources/AdditionalResources.cs
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Puxbay.SDK.Models;

namespace Puxbay.SDK.Resources
{
    /// <summary>
    /// Inventory resource - Inventory tracking operations.
    /// </summary>
    public class InventoryResource : BaseResource
    {
        public InventoryResource(PuxbayClient client) : base(client) { }
        
        public async Task<List<StockLevel>> GetStockLevelsAsync(string branchId)
        {
            return await Client.RequestAsync<List<StockLevel>>(
                HttpMethod.Get, 
                $"inventory/stock-levels/?branch={branchId}"
            );
        }
        
        public async Task<StockLevel> GetProductStockAsync(string productId, string branchId)
        {
            return await Client.RequestAsync<StockLevel>(
                HttpMethod.Get, 
                $"inventory/product-stock/?product={productId}&branch={branchId}"
            );
        }
    }
    
    /// <summary>
    /// Reports resource - Sales reports and analytics.
    /// </summary>
    public class ReportsResource : BaseResource
    {
        public ReportsResource(PuxbayClient client) : base(client) { }
        
        public async Task<FinancialSummary> GetFinancialSummaryAsync(string startDate, string endDate)
        {
            return await Client.RequestAsync<FinancialSummary>(
                HttpMethod.Get, 
                $"reports/financial-summary/?start_date={startDate}&end_date={endDate}"
            );
        }
        
        public async Task<List<DailySales>> GetDailySalesAsync(string startDate, string endDate)
        {
            return await Client.RequestAsync<List<DailySales>>(
                HttpMethod.Get, 
                $"reports/daily-sales/?start_date={startDate}&end_date={endDate}"
            );
        }
        
        public async Task<List<Product>> GetTopProductsAsync(int limit = 10)
        {
            return await Client.RequestAsync<List<Product>>(
                HttpMethod.Get, 
                $"reports/top-products/?limit={limit}"
            );
        }
        
        public async Task<List<Product>> GetLowStockAsync()
        {
            return await Client.RequestAsync<List<Product>>(HttpMethod.Get, "reports/low-stock/");
        }
    }
    
    /// <summary>
    /// Purchase Orders resource - Purchase order management.
    /// </summary>
    public class PurchaseOrdersResource : BaseResource
    {
        public PurchaseOrdersResource(PuxbayClient client) : base(client) { }
        
        public async Task<PaginatedResponse<PurchaseOrder>> ListAsync(int page = 1, string status = null)
        {
            var endpoint = $"purchase-orders/?page={page}";
            if (!string.IsNullOrEmpty(status))
            {
                endpoint += $"&status={status}";
            }
            return await Client.RequestAsync<PaginatedResponse<PurchaseOrder>>(HttpMethod.Get, endpoint);
        }
        
        public async Task<PurchaseOrder> GetAsync(string poId)
        {
            return await Client.RequestAsync<PurchaseOrder>(HttpMethod.Get, $"purchase-orders/{poId}/");
        }
        
        public async Task<PurchaseOrder> CreateAsync(PurchaseOrder po)
        {
            return await Client.RequestAsync<PurchaseOrder>(HttpMethod.Post, "purchase-orders/", po);
        }
        
        public async Task<PurchaseOrder> UpdateAsync(string poId, PurchaseOrder po)
        {
            return await Client.RequestAsync<PurchaseOrder>(HttpMethod.Patch, $"purchase-orders/{poId}/", po);
        }
        
        public async Task<PurchaseOrder> ReceiveAsync(string poId, List<PurchaseOrderItem> items)
        {
            var body = new { items };
            return await Client.RequestAsync<PurchaseOrder>(HttpMethod.Post, $"purchase-orders/{poId}/receive/", body);
        }
    }
    
    /// <summary>
    /// Stock Transfers resource - Stock transfer management.
    /// </summary>
    public class StockTransfersResource : BaseResource
    {
        public StockTransfersResource(PuxbayClient client) : base(client) { }
        
        public async Task<PaginatedResponse<StockTransfer>> ListAsync(int page = 1, string status = null)
        {
            var endpoint = $"stock-transfers/?page={page}";
            if (!string.IsNullOrEmpty(status))
            {
                endpoint += $"&status={status}";
            }
            return await Client.RequestAsync<PaginatedResponse<StockTransfer>>(HttpMethod.Get, endpoint);
        }
        
        public async Task<StockTransfer> GetAsync(string transferId)
        {
            return await Client.RequestAsync<StockTransfer>(HttpMethod.Get, $"stock-transfers/{transferId}/");
        }
        
        public async Task<StockTransfer> CreateAsync(StockTransfer transfer)
        {
            return await Client.RequestAsync<StockTransfer>(HttpMethod.Post, "stock-transfers/", transfer);
        }
        
        public async Task<StockTransfer> CompleteAsync(string transferId)
        {
            return await Client.RequestAsync<StockTransfer>(HttpMethod.Post, $"stock-transfers/{transferId}/complete/");
        }
    }
    
    /// <summary>
    /// Stocktakes resource - Stocktake session management.
    /// </summary>
    public class StocktakesResource : BaseResource
    {
        public StocktakesResource(PuxbayClient client) : base(client) { }
        
        public async Task<PaginatedResponse<StocktakeSession>> ListAsync(int page = 1)
        {
            return await Client.RequestAsync<PaginatedResponse<StocktakeSession>>(
                HttpMethod.Get, 
                $"stocktakes/?page={page}"
            );
        }
        
        public async Task<StocktakeSession> GetAsync(string stocktakeId)
        {
            return await Client.RequestAsync<StocktakeSession>(HttpMethod.Get, $"stocktakes/{stocktakeId}/");
        }
        
        public async Task<StocktakeSession> CreateAsync(StocktakeSession stocktake)
        {
            return await Client.RequestAsync<StocktakeSession>(HttpMethod.Post, "stocktakes/", stocktake);
        }
        
        public async Task<StocktakeSession> CompleteAsync(string stocktakeId)
        {
            return await Client.RequestAsync<StocktakeSession>(HttpMethod.Post, $"stocktakes/{stocktakeId}/complete/");
        }
    }
    
    /// <summary>
    /// Cash Drawers resource - Cash drawer session management.
    /// </summary>
    public class CashDrawersResource : BaseResource
    {
        public CashDrawersResource(PuxbayClient client) : base(client) { }
        
        public async Task<PaginatedResponse<CashDrawerSession>> ListAsync(int page = 1)
        {
            return await Client.RequestAsync<PaginatedResponse<CashDrawerSession>>(
                HttpMethod.Get, 
                $"cash-drawers/?page={page}"
            );
        }
        
        public async Task<CashDrawerSession> GetAsync(string drawerId)
        {
            return await Client.RequestAsync<CashDrawerSession>(HttpMethod.Get, $"cash-drawers/{drawerId}/");
        }
        
        public async Task<CashDrawerSession> OpenAsync(CashDrawerSession drawer)
        {
            return await Client.RequestAsync<CashDrawerSession>(HttpMethod.Post, "cash-drawers/", drawer);
        }
        
        public async Task<CashDrawerSession> CloseAsync(string drawerId, decimal actualCash)
        {
            var body = new { actual_cash = actualCash };
            return await Client.RequestAsync<CashDrawerSession>(HttpMethod.Post, $"cash-drawers/{drawerId}/close/", body);
        }
    }
    
    /// <summary>
    /// Expenses resource - Expense tracking operations.
    /// </summary>
    public class ExpensesResource : BaseResource
    {
        public ExpensesResource(PuxbayClient client) : base(client) { }
        
        public async Task<PaginatedResponse<Expense>> ListAsync(int page = 1, string category = null)
        {
            var endpoint = $"expenses/?page={page}";
            if (!string.IsNullOrEmpty(category))
            {
                endpoint += $"&category={category}";
            }
            return await Client.RequestAsync<PaginatedResponse<Expense>>(HttpMethod.Get, endpoint);
        }
        
        public async Task<Expense> GetAsync(string expenseId)
        {
            return await Client.RequestAsync<Expense>(HttpMethod.Get, $"expenses/{expenseId}/");
        }
        
        public async Task<Expense> CreateAsync(Expense expense)
        {
            return await Client.RequestAsync<Expense>(HttpMethod.Post, "expenses/", expense);
        }
        
        public async Task<Expense> UpdateAsync(string expenseId, Expense expense)
        {
            return await Client.RequestAsync<Expense>(HttpMethod.Patch, $"expenses/{expenseId}/", expense);
        }
        
        public async Task DeleteAsync(string expenseId)
        {
            await Client.RequestAsync<object>(HttpMethod.Delete, $"expenses/{expenseId}/");
        }
        
        public async Task<List<ExpenseCategory>> ListCategoriesAsync()
        {
            return await Client.RequestAsync<List<ExpenseCategory>>(HttpMethod.Get, "expense-categories/");
        }
    }
    
    /// <summary>
    /// Notifications resource - Notification management.
    /// </summary>
    public class NotificationsResource : BaseResource
    {
        public NotificationsResource(PuxbayClient client) : base(client) { }
        
        public async Task<PaginatedResponse<Notification>> ListAsync(int page = 1)
        {
            return await Client.RequestAsync<PaginatedResponse<Notification>>(
                HttpMethod.Get, 
                $"notifications/?page={page}"
            );
        }
        
        public async Task<Notification> GetAsync(string notificationId)
        {
            return await Client.RequestAsync<Notification>(HttpMethod.Get, $"notifications/{notificationId}/");
        }
        
        public async Task<Notification> MarkAsReadAsync(string notificationId)
        {
            return await Client.RequestAsync<Notification>(HttpMethod.Post, $"notifications/{notificationId}/mark-read/");
        }
    }
    
    /// <summary>
    /// Returns resource - Return processing operations.
    /// </summary>
    public class ReturnsResource : BaseResource
    {
        public ReturnsResource(PuxbayClient client) : base(client) { }
        
        public async Task<PaginatedResponse<Return>> ListAsync(int page = 1)
        {
            return await Client.RequestAsync<PaginatedResponse<Return>>(HttpMethod.Get, $"returns/?page={page}");
        }
        
        public async Task<Return> GetAsync(string returnId)
        {
            return await Client.RequestAsync<Return>(HttpMethod.Get, $"returns/{returnId}/");
        }
        
        public async Task<Return> CreateAsync(Return returnObj)
        {
            return await Client.RequestAsync<Return>(HttpMethod.Post, "returns/", returnObj);
        }
        
        public async Task<Return> ApproveAsync(string returnId)
        {
            return await Client.RequestAsync<Return>(HttpMethod.Post, $"returns/{returnId}/approve/");
        }
    }
}
